server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java

```diff
@@ -2018,4 +2018,33 @@ sub int as round((random() + 1) * 100)
             assertThat(response.rows()[0][0]).isEqualTo(generated);
         }
     }
+
+    @Test
+    public void test_insert_on_conflict_with_non_deterministic_column() {
+        execute("""
+            CREATE TABLE tbl (
+               id int PRIMARY KEY,
+               value DOUBLE PRECISION,
+               value_text TEXT,
+               rnd_col int GENERATED ALWAYS AS random() + 10
+            )""");
+
+        ensureGreen();
+        execute("INSERT INTO tbl (id, value) " +
+            "VALUES (1, 99999) " +
+            "ON CONFLICT (id) DO UPDATE SET value = excluded.value");
+        refresh();
+
+        execute("SELECT underreplicated_shards FROM sys.health WHERE table_name = 'tbl'");
+        assertThat(response).hasRows("0"); // Used to be > 0 because of class cast exception caused by column->value mismatch in the request
+
+        // Ensure that primary and replica has the same values for the generated column.
+        execute("SELECT _raw FROM tbl");
+        String rnd = (String) response.rows()[0][0];
+
+        for (int i = 0; i < 10; i++) {
+            execute("SELECT _raw FROM tbl");
+            assertThat(response.rows()[0][0]).isEqualTo(rnd);
+        }
```

Comment:
Acutally, it's not guaranteed to hit the replica. Can we add a `TransportShardUpsertActionTest` test instead?
Also I think the related bug would appear already while writing the row, so the read tests aren't related afaik.

Reply:
Added. Some details: neither `processRequestItems` nor `processRequestItemsonReplica` which we use in other tests not really fit to expose the bug:

`processRequestItems` is okay on master and `processRequestItemsonReplica` is not even called. 
So bug happens in some intermediate logic which requires mocking almost the whole infrastructure of replication which is not related to `TransportShardUpsertAction`.

What I did is made `processRequestItems` to return successfull result - as on master and got replica request from it. Then simply tested streaming - since it's exactly the place where replication trips on master.